velox/expression/tests/FuzzerRunner.h

```diff
@@ -106,16 +106,139 @@ class FuzzerRunner {
     return output;
   }
 
-  static const std::unordered_map<
+  static std::unordered_map<
       std::string,
       std::vector<facebook::velox::exec::FunctionSignaturePtr>>
-      kSpecialForms;
+      specialFormSignatures;
+
+  /// Add supported signatures for the Cast special form for basic types.
+  /// TODO: Add supported Cast signatures to CastTypedExpr and expose them to
+  /// fuzzer instead of hard-coding signatures here.
+  static void addSignaturesForCast() {
```

Comment:
This function is very long and hard to read. Let's move it to .cpp file and refactor.

For example, we can shorten the code by introducing makeCastSignature helper function:

```
  static facebook::velox::exec::FunctionSignaturePtr makeCastSignature(
      const std::string& fromType,
      const std::string& toType) {
    return facebook::velox::exec::FunctionSignatureBuilder()
        .argumentType(fromType)
        .returnType(toType)
        .build();
  }
```

We can also combine int -> float and float -> int loops.

```
  static void addSignaturesForCast() {
    std::vector<std::string> integralTypes{
        "tinyint", "smallint", "integer", "bigint"};
    std::vector<std::string> floatingPointTypes{"real", "double"};

    std::vector<facebook::velox::exec::FunctionSignaturePtr> signatures;

    // integral <--> floating-point.
    for (const auto& fromType : integralTypes) {
      for (const auto& toType : floatingPointTypes) {
        signatures.push_back(makeCastSignature(fromType, toType));
        signatures.push_back(makeCastSignature(fromType, toType));
      }
    }

    // integral <--> varchar.
    for (const auto& type : integralTypes) {
      signatures.push_back(makeCastSignature(type, "varchar"));
      signatures.push_back(makeCastSignature("varchar", type));
    }

    // floating-point <--> varchar.
    for (const auto& type : floatingPointTypes) {
      signatures.push_back(makeCastSignature(type, "varchar"));
      signatures.push_back(makeCastSignature("varchar", type));
    }

    // floating-point --> floating-point.
    for (const auto& fromType : floatingPointTypes) {
      for (const auto& toType : floatingPointTypes) {
        if (fromType != toType) {
          signatures.push_back(makeCastSignature(fromType, toType));
        }
      }
    }

    signatures.push_back(makeCastSignature("boolean", "varchar"));
    signatures.push_back(makeCastSignature("varchar", "timestamp"));
    signatures.push_back(makeCastSignature("varchar", "date"));
    signatures.push_back(makeCastSignature("date", "timestamp"));
    signatures.push_back(makeCastSignature("timestamp", "date"));

    for (const auto& basicSignature : signatures) {
      const auto& baseName = basicSignature->argumentTypes()[0].baseName();
      const auto& returnName = basicSignature->returnType().baseName();

      auto fromType = fmt::format("array({})", baseName);
      auto toType = fmt::format("array({})", returnName);
      signatures.push_back(makeCastSignature(fromType, toType));

      fromType = fmt::format("map(varchar, {})", baseName);
      toType = fmt::format("map(varchar, {})", returnName);
      signatures.push_back(makeCastSignature(fromType, toType));

      fromType = fmt::format("row({})", baseName);
      toType = fmt::format("row({})", returnName);
      signatures.push_back(makeCastSignature(fromType, toType));
    }

    specialFormSignatures.emplace("cast", signatures);
  }
```

Reply:
Good suggestions. I've extracted common code into helper functions and shortened this function. 

With regards to how the supported conversions are determined, I couldn't find any information about supported conversions in either Presto's or Velox's documentation. So I looked at the unit tests in CastExprTest.cpp. I also looked at the implementation of conversion kernels used in CastExpr (in Conversions.h) since some supported conversions are not covered by the unit tests.